warehouse/integrations/snowflake/snowflake.go

```diff
@@ -1446,8 +1460,179 @@ func (sf *Snowflake) LoadTestTable(
 		sf.authString(),
 	)
 
-	_, err = sf.DB.ExecContext(ctx, sqlStatement)
-	return
+	_, err := sf.DB.ExecContext(ctx, sqlStatement)
+	return err
+}
+
+func (sf *Snowflake) TestFetchSchema(ctx context.Context) error {
+	if !sf.config.privileges.fetchSchema.enabled {
+		_, err := sf.FetchSchema(ctx)
+		return err
+	}
+
+	roles, err := sf.getRoles(ctx)
+	if err != nil {
+		return fmt.Errorf("getting roles: %w", err)
+	}
+	rolesSet := lo.SliceToMap(roles, func(item string) (string, bool) {
+		return item, true
+	})
+
+	schemaIdentifier := sf.schemaIdentifier()
+	showGrantsStatement := fmt.Sprintf(`SHOW GRANTS ON SCHEMA %s;`, schemaIdentifier)
+	privileges, err := sf.getShowGrantsPrivileges(ctx, showGrantsStatement)
+	if err != nil {
+		return fmt.Errorf("getting show grants privileges: %w", err)
+	}
+	filteredPrivilegesMap := lo.FilterSliceToMap(privileges, func(item privilegeGrant) (string, bool, bool) {
+		if _, ok := rolesSet[item.granteeName]; !ok {
+			return "", false, false
+		}
+		return item.privilege, true, true
+	})
+	if hasOwnershipPrivileges := filteredPrivilegesMap["OWNERSHIP"]; hasOwnershipPrivileges {
+		return nil
+	}
+	missingPrivileges := lo.Filter(sf.config.privileges.fetchSchema.required, func(privilege string, index int) bool {
+		return !filteredPrivilegesMap[privilege]
+	})
+	if len(missingPrivileges) > 0 {
+		return fmt.Errorf("missing privileges: %v on schema %s", missingPrivileges, schemaIdentifier)
+	}
+	return nil
+}
+
+func (sf *Snowflake) getRoles(ctx context.Context) ([]string, error) {
+	role := sf.Warehouse.GetStringDestinationConfig(sf.conf, model.RoleSetting)
+	if len(role) != 0 {
+		return []string{role}, nil
+	}
+	return sf.getGrantedRoles(ctx)
+}
+
+func (sf *Snowflake) getGrantedRoles(ctx context.Context) ([]string, error) {
+	user := sf.Warehouse.GetStringDestinationConfig(sf.conf, model.UserSetting)
+	sqlStatement := fmt.Sprintf("SHOW GRANTS TO USER %q;", user)
```

Comment:
If role is left out on destination config page and snowflake user has a default role set, snowflake uses the default role for that session. Switching to `SHOW GRANTS TO USER` shows all the roles. current logic is checking privileges granted for all the roles. Is this something to check?

Reply:
https://github.com/rudderlabs/rudder-server/pull/5692#discussion_r2028297507

Reply:
> If role is left out on destination config page and snowflake user has a default role set, snowflake uses the default role for that session.

It's possible that no default role is set as well.